services/federator/src/Federator/Validation.hs

```diff
@@ -60,3 +63,30 @@ validateDomain unparsedDomain = do
 
 throwInward :: Members '[Polysemy.Error InwardError] r => InwardErrorType -> Text -> Sem r a
 throwInward errType errMsg = Polysemy.throw $ InwardError errType errMsg
+
+-- | Normalize the path, and ensure the path begins with "federation/" after normalization
+sanitizePath :: Members '[Polysemy.Error InwardError] r => ByteString -> Sem r ByteString
+sanitizePath originalPath = do
+  -- we parse the path using the URI.ByteString so we can make use of that library's normalization functions
+  uriRef <- case URIB.parseRelativeRef URIB.strictURIParserOptions originalPath of
+    Left err -> throwInward IInvalidEndpoint (cs $ show err)
+    Right ref -> pure ref
+  -- Perform these normalizations:
+  -- - hTtP -> http
+  -- - eXaMpLe.org -> example.org
+  -- - If the scheme is known and the port is the default (e.g. 80 for http) it is removed.
+  -- - If the path is empty, set it to /
+  -- - Rewrite path from /foo//bar///baz to /foo/bar/baz
+  -- - Sorts parameters by parameter name
+  -- - Remove dot segments as per RFC3986 Section 5.2.4
+  let normalized = URIB.normalizeURIRef' URIB.aggressiveNormalization uriRef
+  -- next, we use the Network.URI library to get path segments after normalization,
+  -- and ensure the first segment in the path is "federation"
+  uri <- case parseURIReference (cs normalized) of
```

Comment:
Here you are converting the parsed URI back to a `String`, then reparsing it. Wouldn't it be better to just take the path (as `ByteString`) and directly check that it begins with (ascii-encoded) "federation/"?

Reply:
Done.